accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+
+For example, consider this **hypothetical API** that has a `JsonNumber` class (and not the proposal of using just `JsonNode`)
+```cs
+// Ask for a number with a JSON number token type.
+// Internally probably stored as a byte[] or JsonElement.
+JsonNumber number = JsonSerializer.Deserialize<JsonNumber>("1.2");
+double dlb = number.GetValue<double>();
+
+// Change to "NaN".
+number.Value = double.Nan;
+string json = number.Serialize();
+Debug.Assert(json == "\"NaN\""); // Due to quoted number support, this works.
+
+// However what about round-tripping?
+// Can a JSON string be deserialized into a JsonNumber? If so, "JsonNumber" no longer means a "JSON number".
+// When deserializing a string there is no known CLR type, so the internal value would need be based on a string\byte[].
+// So a "JsonString" node would normally be created, not "JsonNumber" and thus throw here:
+number = JsonSerializer.Deserialize<JsonNumber>(json); // Throws
+
+// Even if the above could be made to work, it wouldn't work with a JsonObject node.
+// A JsonObject would create "JsonString" nodes for "NaN" JSON strings since it has
+// no way of knowing these should map to a JsonNumber (or a CLR number).
+```
+
+Thus the DOM is based on CLR types (not JSON tokens) in edit mode and during serialization. The JSON "value" types (number, string, boolean) are represented as a common `JsonValue` sealed class, and **not** separate `JsonNumber`, `JsonString` and `JsonBoolean` classes as shown in the example above.
+
+During serialization, the DOM is written using the serializer and uses any custom converters that may be configured. In all but idiocentric scenarios (e.g. a custom converter can do anything), the behavior is natural and JSON is round-tripped as expected.
+
+In the example, a JSON "NaN" string is deserialized into a `JsonValue` and when `GetValue<double>` is called, a `double` is returned, assuming the appropriate `JsonSerializerOptions` are configured for quoted numbers. When the node is serialized, it produces a JSON string (again, if configured). The consumer of the DOM is not necessarily aware that `double` values assigned to the node may be serialized to JSON as either a string or a JSON number, and after deserialization is not aware what the internal state is - it could be a `byte[]`, a `JsonElement` or a `string` depending on implementation and\or the JSON token (string or number).
+
+## Primary API
+All types live in STJ.dll.
+```cs
+namespace System.Text.Json
+// Also consider existing S.T.J.Serialization namespace instead of S.T.J.
+// since it is now coupled to the serializer rather than stand-alone.
+{
+    public abstract class JsonNode
+    {
+        internal JsonNode(); // prevent external derived classes.
+        public JsonNode(JsonSerializerOptions options = null);
+
+        public JsonSerializerOptions Options { get; }
```

Comment:
Given we have `JsonDocument` and `JsonDocumentOptions`, its odd to see the use of `JsonSerializerOptions` here. Why is this required? Can we avoid the tight coupling, and for anyone wanting to serialize a writable dom into a C# POCO type, I would expect a JsonSerializer overload that accepts it, identical to these: https://github.com/dotnet/runtime/issues/31274

This fits the layering mental model a bit more.

Reply:
Supporting the `JsonSerializerOptions` is covered in the design to interop with various features including:
- `Converters.Add()` (custom converters added at run-time)
- `NumberHandling`
- `DefaultIgnoreCondition` (default\null handling )
- `PropertyNamingPolicy`.  Note that `DictionaryKeyPolicy` is not used since an object and a dictionary both have the same JSON representation.
- `PropertyNameCaseInsensitive`
- `AllowTrailingCommas`
- `ReadCommentHandling`. Note that round-tripping comments are not possible. See the "Features not proposed in first round" section.
- `MaxDepth`

`JsonDocument` \ `JsonElement` are not required for the writable DOM. We could still add those `JsonSerializer.Serialize`\`Deserialize` overloads to the serializer, however, independently.